scripts/f5-config.py

```diff
@@ -14,31 +14,57 @@
 # limitations under the License.
 #
 # (c) 2014, Kevin Carter <kevin.carter@rackspace.com>
+from __future__ import print_function
 import argparse
 import json
 import os
+import sys
 
 import netaddr
 
+IS_LEM = True
 
-PART = 'RPC'
-PREFIX_NAME = 'RPC'
+if 'LAB_DATA_DIR' in os.environ:
+    # If we have LAD_DATA_DIR set, we are operating inside a lab environment
+    LAB_DATA_DIR = os.environ['LAB_DATA_DIR']
+else:
+    print('Not a LEM install, missing LAB_DATA_DIR')
+    IS_LEM = False
+
+
+if 'LAB_NAME' in os.environ:
+    LAB_NAME = os.environ['LAB_NAME'].lower()
+    LAB_NAME_URL = LAB_NAME + ".rpc.rackspace.com"
+else:
+    print('Not a LEM install, missing LAB_NAME')
+    IS_LEM = False
+    LAB_NAME_URL = None
+
+if IS_LEM is True:
+    PART = LAB_NAME.upper()
+    PREFIX_NAME = LAB_NAME.upper()
+else:
+    PART = 'RPC'
+    PREFIX_NAME = 'RPC'
 
 SNAT_POOL = (
     '### CREATE SNATPOOL ###\n'
     'create ltm snatpool /' + PART + '/' + PREFIX_NAME + '_SNATPOOL { members replace-all-with {'
     ' %(snat_pool_addresses)s } }'
 )
 
-#Persistance Profile:
-PERSISTANCE = [
+#  Persistence Profile:
+PERSISTENCE = [
     r'create ltm persistence source-addr /' + PART + '/' + PREFIX_NAME + '_PROF_PERSIST_IP {'
     r' app-service none defaults-from /Common/source_addr'
     r' match-across-services enabled timeout 3600 }',
     r'create ltm persistence cookie /' + PART + '/' + PREFIX_NAME + '_PROF_PERSIST_COOKIE {'
     r' app-service none cookie-name RPC-COOKIE defaults-from /Common/cookie }''\n'
 ]
 
+
+#  TODO: Configure monitoring user
```

Comment:
Configure Monitoring user is adding a user into the openstack environment with access to check endpoints.
The RPC-MON-EXT-ENDPOINT.monitor is a shell script the F5 uses to check the status

<pre>
#!/usr/bin/env bash

# Copy stdout to fd 3 and redirect stdout/stderr to /var/log/f5-monitor.log
exec 3>&1 &> /var/log/f5-monitor.log #/dev/null
#/var/log/f5-monitor.log - replace null with this file to enable logging when executing the script

# Auth connection
auth_proto="http"
auth_ip="INSERT-IP-HERE"
    #Internal IP address of f5 VIP
auth_port="5000"
auth_url="$auth_proto://$auth_ip:$auth_port/v2.0/tokens"

# Auth args pulled from source file
tenant="admin"
user="admin"
pass="INSERTPWHERE"
tenant_id="INSERTIDHERE"

# Save token to file
save_token() {
    echo "$token" > /var/tmp/keystone-token
}

# Get new token
new_token() {
    # Curl new token from keystone from user/pass for tenant
    IFS=$'\n' read -rd '' -a resp < <(curl -sk $auth_url -X POST -H "Accept: application/json" -H "Content-Type: application/json" -H "User-Agent: f5-ltm" -d @- -w "\n%{http_code}" <<EOF
{
    "auth": {
        "tenantName": "$tenant",
        "passwordCredentials": {
            "username": "$user",
            "password": "$pass"
        }
    }
}
EOF
    )

    # Exit if status is 401 (invalid username/pass)
    if [[ "${resp[@]:(-1)}" == "401" ]]
    then
        echo "Exiting after failure to get a valid token: Check user/pass."
        printf "%s\n" "${resp[@]}"
        exit -1
    fi

    # Set token variable
    token=$(sed -nr '/\{"access": \{"token": \{"issued_at": "([-0-9TZ.:]*)", "expires": "([-0-9TZ.:]*)", "id": "([0-9a-z]*)", .*?$/ s//\3/p' <<< "${resp[@]}")

    # Might need to parse tenant ID
    #tenant=...

    # Mark as new token
    token_new=1

    # Save to file
    save_token
}

# Read token from file or get new one
get_token() {
    # If file exists
    if [[ -f /var/tmp/keystone-token ]]
    then
        # Read into token variable
        read -r token < /var/tmp/keystone-token
    else
        # Get new one
        new_token
      fi
  }

  do_check() {
      # Check connection
      check_proto="$auth_proto"
      check_ip=$1
      check_port=$2
      expected_statuses="200"

      # Build check url by port
      case $check_port in
          5000)
              #Keystone
              check_url="$check_proto://$check_ip:$check_port/v2.0/"
              ;;
          8776)
              #Cinder
              check_url="$check_proto://$check_ip:$check_port/v2/"
              ;;
          9292)
              #Glance API
              check_url="$check_proto://$check_ip:$check_port/v1/"
              ;;
          9191)
              #Glance Registry
              check_url="$check_proto://$check_ip:$check_port/"
              ;;
          9696)
              #Neutron Server
              check_url="$check_proto://$check_ip:$check_port/v2.0/"
              ;;
          8774)
              #Nova API Compute
              check_url="$check_proto://$check_ip:$check_port/v3/"
              ;;
          8004)
              #Heat API
              check_url="$check_proto://$check_ip:$check_port/v1/$tenant_id/stacks?"
              ;;
          8080)
              #Swift
              check_url="$check_proto://$check_ip:$check_port/v1/AUTH_$tenant_id"
              expected_statuses="200 204"
              ;;
          *)
              # Guard
              echo "Invalid port specified; bailing out"
              exit -1
      esac

      # Check endpoint
      echo IFS=$'\n' read -rd '' -a resp < <(curl -s -I -X GET -H "User-Agent: f5-ltm" -H "X-Auth-Token: $token"  $check_url -w "\n%{http_code}")
      IFS=$'\n' read -rd '' -a resp < <(curl -s -I -X GET -H "User-Agent: f5-ltm" -H "X-Auth-Token: $token"  $check_url -w "\n%{http_code}")

      # Store status code
      status="${resp[@]:(-1)}"

      # Check if we've got an expected status code
      for status_code in $expected_statuses; do
          if [[ "$status" == "$status_code" ]]
          then
              echo "Success" >&3
              exit 0
          fi
      done

      # Check for 401 (token expiration or unauthorized)
      if [[ "$status" == "401" ]]
      then
          # Exit if token is new
          if [[ "$token_new" == "1" ]]
          then
              echo "Exiting after failure to authorize with valid token $token on $check_url"
              printf "%s\n" "${resp[@]}"
              exit -1
          # Else we tried cached token
          else
              # Get a new token and try again
              new_token
              do_check
          fi
      # Something else happened, so bail
      else
          echo "Exiting on status: $status"
          printf "%s\n" "${resp[@]}"
          exit -1
      fi
  }

  # Get token
  get_token
  # Do endpoint check
  do_check $1 $2
</pre>


This is a chicken and egg issue when setting up a new deployment. As in order to do the deployment the services need to be accessible on the F5. The monitoring user isn't created and the F5 flags the VS as down. 

In the automated lab environment I had no idea how to add the user if the endpoints were not accessible. RPC support adds the external monitor manually after the deployment.

The LEM bypasses this check using the following shell on the output of the F5_config.py

<pre>
#!/usr/bin/env bash
# Just a note, if your running this on a mac, sed sux, use gsed as it's not the same on ubuntu

if [ -z "$LAB_NAME" ]; then
    echo "Need to set LAB_NAME before using."
    exit
fi

UPPER_NAME="${LAB_NAME^^}"

./f5_config.py -f $1 -e ./data/$LAB_NAME.f5

# Strip out the external monitor and use TCP half open for lab environments
sed "s/ monitor \/$UPPER_NAME\/RPC-MON-EXT-ENDPOINT/ monitor \/Common\/tcp_half_open/g" -i ./data/$LAB_NAME.f5

# Lets not use the GALERA monitor either, in lab upgrades, use half open
sed "s/ monitor \/$UPPER_NAME\/${UPPER_NAME}_MON_GALERA/ monitor \/Common\/tcp_half_open/g" -i ./data/$LAB_NAME.f5


./f5_tmsh.py -f ./data/$LAB_NAME.f5

</pre>

Since I really don't care if it passes health checks with tokens, and just a port is available is good enough for an automated lab deployment.

I wrote the shell script/ sed so I would not have to change the output of the F5_config.py script.